File: go/datastructures/datastructures_test.go
package datastructures

import (
	"testing"
)

func TestStack(t *testing.T) {
	stack := NewStack[int]()

	if !stack.IsEmpty() {
		t.Error("New stack should be empty")
	}

	stack.Push(1)
	stack.Push(2)
	stack.Push(3)

	if stack.Size() != 3 {
		t.Errorf("Expected size 3, got %d", stack.Size())
	}

	val, ok := stack.Peek()
	if !ok || val != 3 {
		t.Errorf("Expected peek 3, got %d", val)
	}

	val, ok = stack.Pop()
	if !ok || val != 3 {
		t.Errorf("Expected pop 3, got %d", val)
	}

	val, ok = stack.Pop()
	if !ok || val != 2 {
		t.Errorf("Expected pop 2, got %d", val)
	}
}

func TestQueue(t *testing.T) {
	queue := NewQueue[int]()

	if !queue.IsEmpty() {
		t.Error("New queue should be empty")
	}

	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Enqueue(3)

	if queue.Size() != 3 {
		t.Errorf("Expected size 3, got %d", queue.Size())
	}

	val, ok := queue.Peek()
	if !ok || val != 1 {
		t.Errorf("Expected peek 1, got %d", val)
	}

	val, ok = queue.Dequeue()
	if !ok || val != 1 {
		t.Errorf("Expected dequeue 1, got %d", val)
	}

	val, ok = queue.Dequeue()
	if !ok || val != 2 {
		t.Errorf("Expected dequeue 2, got %d", val)
	}
}

func TestDynamicArray(t *testing.T) {
	arr := NewDynamicArray[int](2)

	if !arr.IsEmpty() {
		t.Error("New array should be empty")
	}

	for i := 0; i < 100; i++ {
		arr.Push(i)
	}

	if arr.Size() != 100 {
		t.Errorf("Expected size 100, got %d", arr.Size())
	}

	val, ok := arr.Get(99)
	if !ok || val != 99 {
		t.Errorf("Expected get(99) = 99, got %d", val)
	}
}

func TestSinglyLinkedList(t *testing.T) {
	list := NewSinglyLinkedList[int]()

	list.PushFront(1)
	list.PushFront(2)
	list.PushBack(3)

	if list.Size() != 3 {
		t.Errorf("Expected size 3, got %d", list.Size())
	}

	front, _ := list.PeekFront()
	back, _ := list.PeekBack()

	if front != 2 {
		t.Errorf("Expected front 2, got %d", front)
	}
	if back != 3 {
		t.Errorf("Expected back 3, got %d", back)
	}

	if !list.Contains(1) {
		t.Error("List should contain 1")
	}
}

func TestDoublyLinkedList(t *testing.T) {
	list := NewDoublyLinkedList[int]()

	list.PushFront(1)
	list.PushFront(2)
	list.PushBack(3)

	if list.Size() != 3 {
		t.Errorf("Expected size 3, got %d", list.Size())
	}

	val, _ := list.PopFront()
	if val != 2 {
		t.Errorf("Expected pop front 2, got %d", val)
	}

	val, _ = list.PopBack()
	if val != 3 {
		t.Errorf("Expected pop back 3, got %d", val)
	}
}

func TestDeque(t *testing.T) {
	deque := NewDeque[int]()

	deque.PushFront(1)
	deque.PushBack(2)
	deque.PushFront(0)

	if deque.Size() != 3 {
		t.Errorf("Expected size 3, got %d", deque.Size())
	}

	front, _ := deque.PeekFront()
	back, _ := deque.PeekBack()

	if front != 0 {
		t.Errorf("Expected front 0, got %d", front)
	}
	if back != 2 {
		t.Errorf("Expected back 2, got %d", back)
	}
}

func TestBinarySearchTree(t *testing.T) {
	bst := NewBinarySearchTree[int]()

	bst.Insert(5)
	bst.Insert(3)
	bst.Insert(7)
	bst.Insert(1)
	bst.Insert(9)

	if bst.Size() != 5 {
		t.Errorf("Expected size 5, got %d", bst.Size())
	}

	if !bst.Contains(3) {
		t.Error("BST should contain 3")
	}
	if bst.Contains(4) {
		t.Error("BST should not contain 4")
	}

	inorder := bst.InOrder()
	expected := []int{1, 3, 5, 7, 9}
	for i, v := range inorder {
		if v != expected[i] {
			t.Errorf("Inorder mismatch at %d: expected %d, got %d", i, expected[i], v)
		}
	}
}

func TestMinHeap(t *testing.T) {
	heap := NewMinHeap[int]()

	heap.Insert(5)
	heap.Insert(3)
	heap.Insert(7)
	heap.Insert(1)

	if heap.Size() != 4 {
		t.Errorf("Expected size 4, got %d", heap.Size())
	}

	min, _ := heap.Peek()
	if min != 1 {
		t.Errorf("Expected peek 1, got %d", min)
	}

	min, _ = heap.ExtractMin()
	if min != 1 {
		t.Errorf("Expected extract 1, got %d", min)
	}

	min, _ = heap.Peek()
	if min != 3 {
		t.Errorf("Expected peek 3, got %d", min)
	}
}

func TestDisjointSet(t *testing.T) {
	ds := NewDisjointSet(5)

	if ds.Connected(0, 1) {
		t.Error("0 and 1 should not be connected initially")
	}

	ds.Union(0, 1)
	if !ds.Connected(0, 1) {
		t.Error("0 and 1 should be connected after union")
	}

	if ds.Count() != 4 {
		t.Errorf("Expected count 4, got %d", ds.Count())
	}

	ds.Union(2, 3)
	ds.Union(0, 2)

	if !ds.Connected(0, 3) {
		t.Error("0 and 3 should be connected")
	}

	if ds.Count() != 2 {
		t.Errorf("Expected count 2, got %d", ds.Count())
	}
}

File: php/tests/DataStructuresTest.php
<?php

declare(strict_types=1);

namespace DSA\Tests;

use PHPUnit\Framework\TestCase;
use DSA\DataStructures\Stack;
use DSA\DataStructures\Queue;
use DSA\DataStructures\DynamicArray;
use DSA\DataStructures\SinglyLinkedList;
use DSA\DataStructures\DoublyLinkedList;
use DSA\DataStructures\Deque;
use DSA\DataStructures\HashTable;
use DSA\DataStructures\BinarySearchTree;
use DSA\DataStructures\MinHeap;
use DSA\DataStructures\DisjointSet;

class DataStructuresTest extends TestCase
{
    // Stack Tests
    public function testStackOperations(): void
    {
        $stack = new Stack();
        $this->assertTrue($stack->isEmpty());

        $stack->push(1);
        $stack->push(2);
        $stack->push(3);

        $this->assertEquals(3, $stack->size());
        $this->assertEquals(3, $stack->peek());
        $this->assertEquals(3, $stack->pop());
        $this->assertEquals(2, $stack->pop());
        $this->assertEquals(1, $stack->size());

        $stack->clear();
        $this->assertTrue($stack->isEmpty());
    }

    // Queue Tests
    public function testQueueOperations(): void
    {
        $queue = new Queue();
        $this->assertTrue($queue->isEmpty());

        $queue->enqueue(1);
        $queue->enqueue(2);
        $queue->enqueue(3);

        $this->assertEquals(3, $queue->size());
        $this->assertEquals(1, $queue->peek());
        $this->assertEquals(1, $queue->dequeue());
        $this->assertEquals(2, $queue->dequeue());
        $this->assertEquals(1, $queue->size());
    }

    // DynamicArray Tests
    public function testDynamicArrayOperations(): void
    {
        $arr = new DynamicArray();
        $this->assertTrue($arr->isEmpty());

        $arr->push(10);
        $arr->push(20);
        $arr->push(30);

        $this->assertEquals(3, $arr->size());
        $this->assertEquals(10, $arr->get(0));
        $this->assertEquals(30, $arr->get(2));

        $arr->set(1, 25);
        $this->assertEquals(25, $arr->get(1));

        $this->assertEquals(30, $arr->pop());
        $this->assertEquals(2, $arr->size());

        $arr->insert(1, 15);
        $this->assertEquals([10, 15, 25], $arr->toArray());

        $arr->removeAt(1);
        $this->assertEquals([10, 25], $arr->toArray());
    }

    // SinglyLinkedList Tests
    public function testSinglyLinkedListOperations(): void
    {
        $list = new SinglyLinkedList();
        $this->assertTrue($list->isEmpty());

        $list->append(1);
        $list->append(2);
        $list->prepend(0);

        $this->assertEquals(3, $list->size());
        $this->assertEquals(0, $list->get(0));
        $this->assertEquals(2, $list->get(2));

        $list->insert(1, 10);
        $this->assertEquals([0, 10, 1, 2], $list->toArray());

        $list->removeAt(1);
        $this->assertEquals([0, 1, 2], $list->toArray());

        $this->assertTrue($list->contains(1));
        $this->assertFalse($list->contains(100));
    }

    // DoublyLinkedList Tests
    public function testDoublyLinkedListOperations(): void
    {
        $list = new DoublyLinkedList();
        $this->assertTrue($list->isEmpty());

        $list->append(1);
        $list->append(2);
        $list->prepend(0);

        $this->assertEquals(3, $list->size());
        $this->assertEquals(0, $list->getFirst());
        $this->assertEquals(2, $list->getLast());

        $this->assertEquals(0, $list->removeFirst());
        $this->assertEquals(2, $list->removeLast());
        $this->assertEquals(1, $list->size());
    }

    // Deque Tests
    public function testDequeOperations(): void
    {
        $deque = new Deque();
        $this->assertTrue($deque->isEmpty());

        $deque->pushBack(2);
        $deque->pushFront(1);
        $deque->pushBack(3);

        $this->assertEquals(3, $deque->size());
        $this->assertEquals(1, $deque->peekFront());
        $this->assertEquals(3, $deque->peekBack());

        $this->assertEquals(1, $deque->popFront());
        $this->assertEquals(3, $deque->popBack());
        $this->assertEquals(1, $deque->size());
    }

    // HashTable Tests
    public function testHashTableOperations(): void
    {
        $table = new HashTable();
        $this->assertTrue($table->isEmpty());

        $table->put('one', 1);
        $table->put('two', 2);
        $table->put('three', 3);

        $this->assertEquals(3, $table->size());
        $this->assertEquals(1, $table->get('one'));
        $this->assertEquals(2, $table->get('two'));
        $this->assertNull($table->get('four'));

        $this->assertTrue($table->contains('one'));
        $this->assertFalse($table->contains('four'));

        $table->put('one', 100);
        $this->assertEquals(100, $table->get('one'));

        $this->assertEquals(2, $table->remove('two'));
        $this->assertEquals(2, $table->size());
        $this->assertFalse($table->contains('two'));
    }

    // BinarySearchTree Tests
    public function testBinarySearchTreeOperations(): void
    {
        $bst = new BinarySearchTree();
        $this->assertTrue($bst->isEmpty());

        $bst->insert(5);
        $bst->insert(3);
        $bst->insert(7);
        $bst->insert(1);
        $bst->insert(9);

        $this->assertEquals(5, $bst->size());
        $this->assertTrue($bst->contains(5));
        $this->assertTrue($bst->contains(3));
        $this->assertFalse($bst->contains(100));

        $this->assertEquals(1, $bst->min());
        $this->assertEquals(9, $bst->max());

        $this->assertEquals([1, 3, 5, 7, 9], $bst->inOrder());

        $this->assertTrue($bst->remove(3));
        $this->assertFalse($bst->contains(3));
        $this->assertEquals([1, 5, 7, 9], $bst->inOrder());
    }

    // MinHeap Tests
    public function testMinHeapOperations(): void
    {
        $heap = new MinHeap();
        $this->assertTrue($heap->isEmpty());

        $heap->insert(5);
        $heap->insert(3);
        $heap->insert(7);
        $heap->insert(1);
        $heap->insert(9);

        $this->assertEquals(5, $heap->size());
        $this->assertEquals(1, $heap->peek());

        $this->assertEquals(1, $heap->extractMin());
        $this->assertEquals(3, $heap->extractMin());
        $this->assertEquals(5, $heap->extractMin());
        $this->assertEquals(7, $heap->extractMin());
        $this->assertEquals(9, $heap->extractMin());

        $this->assertTrue($heap->isEmpty());
    }

    public function testMinHeapWithCustomComparator(): void
    {
        // Max heap using custom comparator
        $heap = new MinHeap(fn($a, $b) => $b <=> $a);

        $heap->insert(5);
        $heap->insert(3);
        $heap->insert(7);

        $this->assertEquals(7, $heap->extractMin());
        $this->assertEquals(5, $heap->extractMin());
        $this->assertEquals(3, $heap->extractMin());
    }

    // DisjointSet Tests
    public function testDisjointSetOperations(): void
    {
        $ds = new DisjointSet(5);

        $this->assertEquals(5, $ds->count());
        $this->assertFalse($ds->connected(0, 1));

        $ds->union(0, 1);
        $this->assertTrue($ds->connected(0, 1));
        $this->assertEquals(4, $ds->count());

        $ds->union(2, 3);
        $ds->union(0, 2);

        $this->assertTrue($ds->connected(0, 3));
        $this->assertTrue($ds->connected(1, 2));
        $this->assertEquals(2, $ds->count());

        $this->assertFalse($ds->connected(0, 4));
    }
}
